Blocks: Fix incorrect placement for hooked blocks in the parent container

gziolo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #54307.

Every block that has inner blocks also has some wrapping HTML element. The way hooked blocks are currently implemented is that it puts the injected block at exactly first (`firstChild`) or last place (`lastChild`) in the parsed block, which happens to be outside the wrapping element. A simplified example based on REST API response that shows the issue:

```html
<!-- wp:group {"tagName":"main","style":{"spacing":{"margin":{"top":"var:preset|spacing|50","bottom":"var:preset|spacing|70"}}},"layout":{"type":"constrained"}} -->
    <main class="wp-block-group" style="margin-top:var(--wp--preset--spacing--50);margin-bottom:var(--wp--preset--spacing--70)">
        <!-- wp:pattern {"slug":"twentytwentythree/cta"} /-->
    </main>
    <!-- wp:ockham/like-button /-->
<!-- /wp:group /-->
```

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The injected block is misplaced and therefore incorrectly styled on the front end. It should be with the `main` tag instead in the example provided:

```html
<!-- wp:group {"tagName":"main","style":{"spacing":{"margin":{"top":"var:preset|spacing|50","bottom":"var:preset|spacing|70"}}},"layout":{"type":"constrained"}} -->
    <main class="wp-block-group" style="margin-top:var(--wp--preset--spacing--50);margin-bottom:var(--wp--preset--spacing--70)">
        <!-- wp:pattern {"slug":"twentytwentythree/cta"} /-->
        <!-- wp:ockham/like-button /-->
    </main>
<!-- /wp:group /-->
```

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Instead of injecting the hooked block as the first item (`firstChild`) or the last item (`lastChild`), there is no logic that tries to locate the exact place where the child block is marked with `null` in the array containing `innerContent` of the block. Once the first `null` value is found, then another `null` value is placed next to it. In the case, when there is no child block (no `null` value), then the `null` marker is added as the first item (`firstChild`) or the last item (`lastItem`) in the array.

Note: the handling for the case with no inner blocks but the existing HTML wrapper for inner blocks isn't handled perfectly. We can discuss options for improving that separately.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Install and activate the plugin with the hooked block. Example block can be downloaded from https://github.com/ockham/like-button/releases/tag/v0.4.1.
2. Activate Twenty Twenty-Three theme or any other theme that uses `core/comment-template` block.
3. Go to the single post page and see where the Like Button gets injected. Ensure that all templates and template parts used on that page don't have any customizations applied.
4. Check the source of the page and notice that the Like button is placed correctly before the closing tag for the comment row wrapper.

There is now a test coverage presenting how hooked blocks get injected depending on the state of the container block:
- inner blocks present
- no inner blocks present
- no inner blocks and inner content present

## Screenshots or screencast <!-- if applicable -->

<img width="1248" alt="Screenshot 2023-09-11 at 11 56 31" src="https://github.com/WordPress/gutenberg/assets/699132/88c27c8e-bf19-463b-8077-216e91e174b3">

<img width="976" alt="Screenshot 2023-09-11 at 13 22 39" src="https://github.com/WordPress/gutenberg/assets/699132/69a2b100-005f-43bf-9184-491292b6f591">

gziolo committed:
Blocks: Fix incorrect placement for hooked blocks in the parent container

gziolo committed:
Improve the way chunk index gets detected

gziolo committed:
Add unit tests covering hooking blocks at first and last child positions

gziolo commented:
@ockham, should we add unit tests directly to WordPress core and only apply the fix to the logic in the Gutenberg plugin?

Well, if we are going to keep `gutenberg_serialize_blocks` then maybe it makes sense to add the tests here, too.

ockham commented on lib/experimental/block-hooks.php:
```diff
@@ -158,14 +158,28 @@ function gutenberg_insert_hooked_block( $inserted_block, $relative_position, $an
-				// when rendering blocks, we also need to prepend a value (`null`, to mark a block
-				// location) to that array.
-				array_push( $block['innerContent'], null );
+				// when rendering blocks, we also need to correctly prepend a value (`null`, to mark a block
+				// location) to that array after HTML content for the inner blocks wrapper.
```
Oops, I had `prepend` here instead of `append` 😅 Also, we need to insert _before_ the wrapping HTML:

```suggestion
				// when rendering blocks, we also need to correctly append a value (`null`, to mark a block
				// location) to that array before the remaining HTML content for the inner blocks wrapper.
```

gziolo replied:
Good catch. I think I would also further clarify that it's about the remaining HTML content.

ockham commented:
> @ockham, should we add unit tests directly to WordPress core and only apply the fix to the logic in the Gutenberg plugin?
> 
> Well, if we are going to keep `gutenberg_serialize_blocks` then maybe it makes sense to add the tests here, too.

I think it makes sense to keep unit tests here, at least for now 👍 (Definitely good for confidence in the bugfix.)

ockham commented on lib/experimental/block-hooks.php:
```diff
@@ -158,14 +158,28 @@ function gutenberg_insert_hooked_block( $inserted_block, $relative_position, $an
-				array_push( $block['innerContent'], null );
+				// when rendering blocks, we also need to correctly prepend a value (`null`, to mark a block
+				// location) to that array after HTML content for the inner blocks wrapper.
+				$chunk_index = count( $block['innerContent'] ) - 1;
+				for ( $index = $chunk_index; $index >= 0; $index-- ) {
+					if ( is_null( $block['innerContent'][ $index ] ) ) {
+						$chunk_index = $index;
+						break;
+					}
+				}
```
Noting that this sets `$chunk_index` to `-1` if `$block['innerContent']` is empty. Which I think is okay, since `array_splice()` also works if the offset arg is negative ([see](https://www.php.net/manual/en/function.array-splice.php)).

gziolo replied:
Good catch. I think I should set `$chunk_index` to 0 initially and only update when `null` is in the array. I now realized that it's possible that the array is empty 😄 

I was studying unit tests for the parser earlier today, some interesting findings for `innerContent`:

https://github.com/WordPress/gutenberg/blob/7652de9ed159d4e9698e432c06b83e56777160c8/packages/block-serialization-spec-parser/shared-tests.js#L241-L289

In particular, the fact that inner block locations always become `null`, so you can have multiple of them. However, if there are no inner blocks then we won't have `null`, and it can even be an empty array 🤔 

I need to study some existing blocks to see how they serialize when there are no inner blocks. Group, Columns, Column, it might get tricky. I'm wondering if at least for core blocks we always enforce a HTML wrapper (I would assume yes, but it isn't a requirement for the parser), and whether we show the wrapper when there are no inner blocks inserter. I hope that with Columns that can be explored.

Basically, what @ockham raised in https://github.com/WordPress/gutenberg/pull/54349#issuecomment-1715869653.

gziolo replied:
An example for the Columns block with 3 columns and the middle one has no child blocks:

```html
<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>First</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Third</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
```

A Group block with no child blocks:

```html
<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"></div>
<!-- /wp:group -->
```

A Group Block with a Cover block where I removed inner blocks:
```html
<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:cover {"url":"http://localhost:8888/wp-content/uploads/2023/09/Maldives.jpg","id":18,"dimRatio":50,"layout":{"type":"constrained"}} -->
<div class="wp-block-cover"><span aria-hidden="true" class="wp-block-cover__background has-background-dim"></span><img class="wp-block-cover__image-background wp-image-18" alt="" src="http://localhost:8888/wp-content/uploads/2023/09/Maldives.jpg" data-object-fit="cover"/><div class="wp-block-cover__inner-container"></div></div>
<!-- /wp:cover --></div>
<!-- /wp:group -->
```

ockham commented:
I wonder if there are any blocks with inner blocks that _don't_ include any block wrapper HTML. It seems unlikely; I guess we'll find out (and can adjust the code accordingly if needed) 😅 

In theory, a hooked block can be inserted as a first or last child into a block that doesn't have any inner blocks at all. This probably doesn't make any sense (and maybe we should forbid it; though that might not be straight-forward); in any case, I think our logic would work even then (and at least not crash 😬 ).

Furthermore, I wonder if there's ever going to be more than one consecutive non-null array element in `innerContent` (since any "plain" HTML is probably just represented as a string), so maybe we wouldn't even need the loop. But it also doesn't hurt to have it 😬

ockham reviewed (approved):
One [documentation nitpick](https://github.com/WordPress/gutenberg/pull/54349#discussion_r1323111305); plus I agree that

> It would be great to land https://github.com/WordPress/gutenberg/pull/54293 first to remove the need for special code for bootstrapping PHP unit tests.

Other than that: :shipit:

gziolo committed:
Update lib/experimental/block-hooks.php

Co-authored-by: Bernie Reiter <96308+ockham@users.noreply.github.com>

gziolo committed:
Small tweaks after code review and rebasing with trunk.
Props to @ockham for spotting issues.

gziolo commented:
I rebased with `trunk` after landing https://github.com/WordPress/gutenberg/pull/54293. I addressed the feedback from @ockham. I plan to add more tests to cover edge cases as discussed in the comments.

gziolo commented on lib/compat/wordpress-6.4/block-hooks.php:
```diff
@@ -158,14 +158,28 @@ function gutenberg_insert_hooked_block( $inserted_block, $relative_position, $an
-				array_push( $block['innerContent'], null );
+				// when rendering blocks, we also need to correctly append a value (`null`, to mark a block
+				// location) to that array before the remaining HTML content for the inner blocks wrapper.
+				$chunk_index = 0;
```
We probably want to set 0 when no elements, and the last index when elements are present:

<img width="382" alt="Screenshot 2023-09-13 at 15 31 58" src="https://github.com/WordPress/gutenberg/assets/699132/6947f1ea-e0c7-47fa-b204-c86b076997d3">

<img width="377" alt="Screenshot 2023-09-13 at 15 32 21" src="https://github.com/WordPress/gutenberg/assets/699132/5030a98d-c414-48c1-94fb-ffac922ca42f">

`null` should be the last element in the array when using `lastChild`. Well, it's hard to tell if there are no other children.

gziolo replied:
I went with a simpler approach `$chunk_index = count( $block['innerContent'] );`. The reason for that is this way, `null` gets appended to the array, rather than prepending. Plus it resolves the original issue.

gziolo committed:
Add tests covering the edge case where the container block does not have inner blocks

gziolo committed:
Improve test coverage for injecting hooked blocks

gziolo commented:
> I wonder if there are any blocks with inner blocks that _don't_ include any block wrapper HTML. It seems unlikely; I guess we'll find out (and can adjust the code accordingly if needed) 😅

As documented in https://github.com/WordPress/gutenberg/pull/54349#discussion_r1325437826, while the block parser allows inner blocks without a wrapper, in practice it doesn't happen because the way inner blocks are implemented in the block editor enforces it with the current design. The block wrapper is necessary for the UI to work correctly, so while it's still possible to not include it, it would most likely make it impossible to select the block without using the List View.

> In theory, a hooked block can be inserted as a first or last child into a block that doesn't have any inner blocks at all. This probably doesn't make any sense (and maybe we should forbid it; though that might not be straight-forward); in any case, I think our logic would work even then (and at least not crash 😬 ).

I added unit tests that document this case when there is also a block wrapper. The way it currently works isn't perfect, but it definitely won't crash. We can discuss later what to do about it.

> Furthermore, I wonder if there's ever going to be more than one consecutive non-null array element in innerContent (since any "plain" HTML is probably just represented as a string), so maybe we wouldn't even need the loop. But it also doesn't hurt to have it 😬

It's possible, see the example for the Columns block in the block fixtures folder from integration tests:

https://github.com/WordPress/gutenberg/blob/999b49c5ba9e07c0991e5b0cbfa2e4ac952fc738/test/integration/fixtures/blocks/core__columns.parsed.json#L32-L38

ockham commented:
>> Furthermore, I wonder if there's ever going to be more than one consecutive non-null array element in innerContent (since any "plain" HTML is probably just represented as a string), so maybe we wouldn't even need the loop. But it also doesn't hurt to have it 😬
>
> It's possible, see the example for the Columns block in the block fixtures folder from integration tests:
>
> https://github.com/WordPress/gutenberg/blob/999b49c5ba9e07c0991e5b0cbfa2e4ac952fc738/test/integration/fixtures/blocks/core__columns.parsed.json#L32-L38

Sorry, I don't see it 😅 In your example, non-null and `null` elements are alternating; but what I meant with consecutive non-null elements is that there are two (or more) non-null elements with no `null`s in betweeen them 🤔 I.e. something like

```json
 "innerContent": [ 
 	"\n\t<div class=\"wp-block-column\">\n\t\t", 
 	"\n\t\t", 
 	null, 
 	"\n\t</div>\n\t" 
 ] 
```

gziolo commented:
> Sorry, I don't see it 😅 In your example, non-null and null elements are alternating; but what I meant with consecutive non-null elements is that there are two (or more) non-null elements with no nulls in betweeen them 🤔

Interesting observation. I wasn't aware it groups HTML into one entry before and one after the child block. Anyway, the loop will run twice in most of the cases when there is a block wrapper, so it's probably fine.
